Python/instrumentation.c

```diff
@@ -332,6 +320,73 @@ _PyInstruction_GetLength(PyCodeObject *code, int offset)
     return 1 + _PyOpcode_Caches[inst.op.code];
 }
 
+static inline uint8_t
+get_original_opcode(_PyCoLineInstrumentationData *line_data, int index)
+{
+    return line_data->data[index*line_data->bytes_per_entry];
+}
+
+static inline uint8_t *
+get_original_opcode_ptr(_PyCoLineInstrumentationData *line_data, int index)
+{
+    return &line_data->data[index*line_data->bytes_per_entry];
+}
+
+static inline void
+set_original_opcode(_PyCoLineInstrumentationData *line_data, int index, uint8_t opcode)
+{
+    line_data->data[index*line_data->bytes_per_entry] = opcode;
+}
+
+static inline int
+get_line_delta(_PyCoLineInstrumentationData *line_data, int index)
+{
+    uint8_t *ptr = &line_data->data[index*line_data->bytes_per_entry+1];
+    uint32_t value = *ptr;
+    assert(line_data->bytes_per_entry >= 2);
+    if (line_data->bytes_per_entry > 2) {
+        ptr++;
+        value = (value << 8) | *ptr;
+        if (line_data->bytes_per_entry > 3) {
+            ptr++;
+            value = (value << 8) | *ptr;
+            if (line_data->bytes_per_entry > 4) {
+                assert(line_data->bytes_per_entry == 5);
+                ptr++;
+                value = (value << 8) | *ptr;
+            }
+        }
+    }
+    assert(value < INT_MAX);
+    /* NO_LINE is stored as zero. */
+    return ((int)value) + NO_LINE;
+}
+
+static inline void
+set_line_delta(_PyCoLineInstrumentationData *line_data, int index, int line_delta)
+{
+    /* Store line_delta + 2 as we need -2 to represent no line number */
+    assert(line_delta >= NO_LINE);
+    uint32_t adjusted = line_delta - NO_LINE;
+    uint8_t *ptr = &line_data->data[index*line_data->bytes_per_entry+1];
+    assert(adjusted < (1ULL << (line_data->bytes_per_entry*8)));
+    assert(line_data->bytes_per_entry >= 2);
+    if (line_data->bytes_per_entry > 2) {
+        if (line_data->bytes_per_entry > 3) {
+            if (line_data->bytes_per_entry > 4) {
+                assert(line_data->bytes_per_entry == 5);
+                *ptr = adjusted >> 24;
```

Comment:
I think we should do a little endian here so we can put this in a loop:

```c
for (int idx = 1; idx < line_data->bytes_per_entry; idx++) {
    *ptr = adjusted & 0xff;
    ptr++;
    adjusted >>= 8;
};
```

Same thing for decoding.

Reply:
A loop does a lot of extra work for the common case of `line_data->bytes_per_entry == 2`, so I'll do the common case first, then loop:
```C
*ptr = adjusted & 0xff;
for (int idx = 2; idx < line_data->bytes_per_entry; idx++) {
    ptr++;
    adjusted >>= 8;
    *ptr = adjusted & 0xff;
};
```

Reply:
Sure that works too. Not sure how much impact it would actually cause.